File: docs/guides/swift/Sources/saveImageClassificationsAndSettings.swift
import Foundation
#if os(Linux) // For linux interop
    import FoundationNetworking
#endif

import Core
import Search

func saveImageClassificationsAndSettings() async throws {
    struct Image: Codable {
        var objectID: String
        var imageURL: String
        var objects: [[String: AnyCodable]]
    }

    func getImageLabels(imageURL _: String, objectID _: String, scoreLimit _: Double) -> Image {
        // Implement your image classification logic here
        Image(objectID: "", imageURL: "", objects: [])
    }

    let client = try SearchClient(appID: "ALGOLIA_APPLICATION_ID", apiKey: "ALGOLIA_API_KEY")

    var records: [Image] = []

    try await client.browseObjects(
        indexName: "<YOUR_INDEX_NAME>",
        browseParams: BrowseParamsObject(),
        aggregator: { (response: BrowseResponse<Image>) in
            records.append(contentsOf: response.hits.map { hit in
                let imageURL = hit.imageURL
                return getImageLabels(imageURL: imageURL, objectID: hit.objectID, scoreLimit: 0.5)
            })
        }
    )

    try await client.partialUpdateObjects(indexName: "<YOUR_INDEX_NAME>", objects: records, createIfNotExists: true)

    var facets: [String] = []
    var attributes: [String] = []

    for record in records {
        for obj in record.objects {
            for (key, values) in obj {
                if values.value is [Any] {
                    facets.append(contentsOf: [
                        "searchable(objects.\(key).label)",
                        "searchable(objects.\(key).score)",
                    ])
                    attributes.append("objects.\(key).label)")
                }
            }
        }
    }

    let currentSettings = try await client.getSettings(indexName: "<YOUR_INDEX_NAME>")

    let settings = IndexSettings(
        attributesForFaceting: (currentSettings.attributesForFaceting ?? []) + facets,
        searchableAttributes: (currentSettings.searchableAttributes ?? []) + attributes
    )

    try await client.setSettings(indexName: "<YOUR_INDEX_NAME>", indexSettings: settings)
}

File: docs/guides/csharp/src/saveImageClassificationsAndSettings.cs
namespace Algolia;

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using Algolia.Search.Clients;
using Algolia.Search.Http;
using Algolia.Search.Models.Search;

class SaveImageClassificationsAndSettings
{
  class Image
  {
    public required string ImageUrl { get; set; }
    public required string ObjectId { get; set; }
    public required List<Dictionary<string, object>> Objects { get; set; }
  }

  // Retrieve labels
  async Task<Image> GetImageLabels(string imageURL, string objectID, float scoreLimit)
  {
    // Implement your image classification logic here
    return await Task.Run(
      () =>
        new Image
        {
          ImageUrl = "",
          ObjectId = "",
          Objects = [],
        }
    );
  }

  async Task Main(string[] args)
  {
    try
    {
      // API key ACL should include editSettings / addObject
      var client = new SearchClient(new SearchConfig("ALGOLIA_APPLICATION_ID", "ALGOLIA_API_KEY"));

      var hits = await client.BrowseObjectsAsync<Image>(
        "<YOUR_INDEX_NAME>",
        new BrowseParamsObject()
      );

      var images = hits.ToList();
      var records = images
        .Select(hit => GetImageLabels(hit.ImageUrl, hit.ObjectId, 0.5f))
        .Select(src => src.Result)
        .ToList();

      // Update records with image classifications
      await client.PartialUpdateObjectsAsync("<YOUR_INDEX_NAME>", records, true);

      List<string> facets = [];
      List<string> attributes = [];

      foreach (var image in images)
      {
        foreach (var obj in image.Objects)
        {
          foreach (var key in obj.Keys)
          {
            if (obj[key] is IEnumerable<object>)
            {
              facets.Add($"searchable(objects.{key}.label)");
              facets.Add($"searchable(objects.{key}.score)");
              attributes.Add($"objects.{key}.label");
            }
          }
        }
      }

      var currentSettings = await client.GetSettingsAsync("<YOUR_INDEX_NAME>");

      var settings = new IndexSettings
      {
        SearchableAttributes = currentSettings.SearchableAttributes.Concat(attributes).ToList(),
        AttributesForFaceting = currentSettings.AttributesForFaceting.Concat(facets).ToList(),
      };

      await client.SetSettingsAsync("<YOUR_INDEX_NAME>", settings);
    }
    catch (Exception e)
    {
      Console.WriteLine(e.Message);
    }
  }
}
